### ISSUE TITLE
Multi-body rolling acts really strange

### ISSUE THREAD

**Issue by MrNex** [MrNex/NGen#21]
When an object is rolling along a surface, and it hits two objects stacked atop each other, things get weird.. I think I need coefficient of rolling resistance to make behavior more stable... More testing is needed.

Warning.. Until this gets fixed-- all physics simulations are going to be pretty.. Ridiculous.
# UPDATE

The problem has in fact been pinpointed. The issue occurs with stacked objects where the object on the bottom is incurring an angular velocity around an axis parallel to the surface. The object goes to rotate, but nothing slows this rotation, so it constantly collides with the floor and the objects stacked above it. This causes it to violently change directions of rotation and eventually explode.

It turns out, rolling resistance will take care of this. Rolling Resistance, by nature, factors in the force an object is exerting normal to a surface, and uses this to determine the proper impedence to incur on the angular motion.

This normal force would be much greater in the case of stacked objects :smiling_imp:. Thus, the angular motion will be more rapidly subsided, and the bottom object will not incur many collisions causing the system to explode. Alternatively, if necessary, the angular motion could overcome this resistance force, in which case the motion will be enough to throw the stacked objects off of the bottom in the proper manner.

The fix will be implemented within a few days. I do not have much time on my hands the next few days :disappointed:.

**Comment by Waqar144** [comment]
I tried this, and the collision doesn't seem all that weird to me unless I enable the `RigidBody_ApplyInstantaneousTorque()` which is commented out. Once this is enabled, even a very slow rolling object i.e sphere upon hitting the cubes, blows them apart. The cube on top goes flying to the right out of the borders.

One other thing that seems out of place to me is that if I increase the rolling speed of the sphere, upon collision the cube goes crazy. It keeps on bouncing for a long time. I don't know, maybe its due to the material. The same happens if you apply a _hard_ rotation on it.

**Comment by MrNex** [comment]
It would be great help if you could be a little bit more specific, the `RigidBody_ApplyInstantaneousTorque()` which you are enabling, are you speaking of line 75 in Implementation.c?

I actually had not tested enabling only that yet.. If that is the problem you may have pointed us in the correct direction to debug to find the solution. If that is the line of code you were speaking about, could you leave it commented out-- and comment back in the block which starts at line 87 of implementation.c.

From there, just comment out `RigidBody_ApplyInstantaneousTorque()` again on line 110. I would be curious to know if the problem is in fact not a multiple body collision problem, but a rolling objects hitting an oriented object problem!

**Edit**
I think I misunderstood you.. Re-reading this, it sounds like this is exactly what you did.. And it did in fact behave correctly? This is good.. What happens when the rolling sphere hits a single block which had an instantaneous torque applied to it previously-- such that it is no longer axis aligned?

If I get to test this before you do I will post the results!

Also, if it is easier to explain, feel free to push changes and link me to the exact Implementation.c file with which you are testing. Send a pull request, even, so I could replicate the simulation on my machine.

**Comment by Waqar144** [comment]
- if you disable all of `RigidBody_ApplyInstantaneousTorque()`. (the one one line 110 is enabled) The collision is quite perfect.
- If you disable all and just enable it on the **sphere** i.e line 148 with v.comp[0] = 1.0f. Things get weird. This one seems to be the culprit.
- If you  enable the one at the last, line 190, things don't get weird unless you increase v[0] > 6 or 7. That's because I think its not a round body.
  I haven't tried the one at line 75 yet.

**Comment by MrNex** [comment]
Very interesting.. It should be noted that Line 148 **or** Line 153 being enabled _should_ have the same effect.. With both enabled it will have a cumulative effect.
With the advent of rolling mechanics, angular momentum now causes a linear impulse-- likewise linear momentum now causes an angular momentum.

So, an object spinning on a surface will begin to move (with presence of friction)
And an object moving on a surface will begin to spin (with presence of friction)

So perhaps line 148 isn't the culprit, but perhaps it is when the net magnitudes of linear or angular momentum get too large- as you pointed out in your 3rd point?

It should also be noted that with rolling mechanics now being present, highly elastic collisions with high frictional coefficients (A `RigidBody` with high `coefficientOfRestitution`, `dynamicFriction`, and `staticFriction` values) encountering a collision while both spinning and moving _will behave very oddly_. Common physical observations no longer hold true (Namely angle of incidence equals the angle of reflection quickly becomes false) and laws of conservation of momentum will _seem_ to break down (Though of course they don't, it only appears that way to the eye). This behavior is outlined [in a paper by Richard L. Garwin](http://www.rpi.edu/dept/phys/Courses/PHYS1150/GarwinSuperBall.pdf).

So be worried about _weird_ behavior meaning explosive results, not necessarily just seemingly _weird_ behavior.

Perhaps the answer really is as simple as introducing Rolling Resistance-- which after all is highly dependent on shape. It is the missing piece of the puzzle.

**Comment by Waqar144** [comment]
Ok, so it seems you are right about line 148 and 153. They have the same effect.

Here's what i think happens. I had the block 87-120 uncommented. So, there were two blocks stacked on top of each other. Apply `RigidBody_ApplyInstantaneousTorque()` and `RigidBody_ApplyImpulse` simultaneously. Obviously, the one with greater magnitude will cause the ball to move in its direction. In this case ApplyImpulse had greater magnitude i.e `-4y` while the had `1.0` at x. So, I think Torque due to `ApplyInstantaneousTorque` gets zeroed out due to force in opposite direction.

But, sometimes the block on the top of stack, upon colison with the sphere, doesn't go in the outward direction i.e it goes towards where the sphere is coming from. I think this is because of the InstantaneousTorque, it doesn't get zero or something.

Or maybe, upon colision, a torque gets applied to lower block which sets it spinning and the sphere is already spinning. Since both of these are spinning it causes the upper block to go off like that.

Line 190, doesn't have to do anything with any of this, since the block is placed separatly

**Comment by MrNex** [comment]
I seemed to have different results.
When I run the simulation (I did not check my average physics timestep length, but I would estimate it to be at about 1-2 ms per step) The ball rolls into the stacked blocks and stops rolling.. It appears everything is working, but the block on the bottom of the stack begins to twitch linearly and angularly (I believe linearly only as a result of the rolling mechanics). It continues this twitching for a while until the block on top falls off, and then the system becomes explosive.

However, your results have given me previously unknown insight.

Here is what I think is happening:
- The rolling sphere collides with the block on the bottom of the stack
- The rolling sphere continues to spin (in place) until friction due to the linear motion of the points of collision on both the block on the bottom of the stack and the sphere cause it to stop spinning.
- As this is happening, the block on the bottom of the stack is receiving angular impulses which are giving it angular momentum opposite the spheres (Due to the rolling mechanics-- this is expected).
- From here, the block on the bottom goes to spin opposite the direction of the sphere, but it collides with the floor, the block on top of itself, and perhaps(?) the sphere.
- This causes an opposite impulse (with some energy lost due to the `CoefficientOfRestitution` not being 1, so the collision is not perfectly elastic) and the block goes to spin in the opposite direction.
  - **However**, the block collided thrice!! With the block above it, the sphere, and the ground due to its spinning motion resulting from the sphere collision. This causes accumulation of impulses. I theorize if the bottom block is given a `coefficientOfRestitution` of `0.3f` or lower (such that it is responsible of the loss of 70% of energy in all collisions it is part of) the system will become stable.
- These collisions keep accumulating (much quicker on slower computers due to larger physics timesteps) until the system explodes.

This would explain why sometimes the block on top goes outward, and other times it doesn't. Because of this Back-And-Forth accumulation of collisions and the variable physics timestep it is up in the air on which direction the bottom block is trying to spin when the system explodes. If the bottom block is spinning the same direction as the sphere the top block would go outward (The direction the sphere was moving) however if the bottom block is spinning the opposite direction when the system explodes, the top block would come towards the camera's initial position.

Would you like to try lowering the `coefficientOfFriction` of the objects to allow for at least 70% loss of energy to adjust for the accumulation of 3 simultaneous collisions?

**Comment by Waqar144** [comment]
Setting the `CoefficientOfRestitution` of 
- bottom block to 0.7
- top to 0.3
- sphere to 0.7

System appears to be quite stable
- bottom block to 0.3
- top to 0.7
- sphere to 0.9

its still ok.

Changing `block->body->dynamicFriction`
- bottom to 0.7
- top to 0.3
- sphere to 0.5

Result: VERY Explosive
(increasing the speed of sphere makes it more stable)

Changing `block->body->dynamicFriction`
- bottom to 1.0
- top to 0.7
- sphere to 0.2

Result: quite stable

the higher the difference(`block->body->dynamicFriction`) between top and bottom, more explosive results

**Comment by MrNex** [comment]
Wow.. That's very strange. The higher the _difference_ is not what I was expecting. Are you sure about that? A change difference of `0.1` seems awfully small to cause that much more explosive results.

`dynamicFriction` would be the variable responsible for causing more or less rolling (versus more or less slipping). These results are very helpful. Thank you for posting them.

I think the conclusion can be drawn from your results that we can have more stable systems if we introduce `coefficientOfRollingResistance` next. It is very wrong that (other than the interference from `intertia`) square blocks have the same capacity of rolling as spheres...

Would you agree?

Also, maybe it is not the difference. Let us take into account that when calculating frictional forces, the average of the two objects `dynamicFriction` coefficients is taken and used in the computation. The second example:
- bottom to 1.0
- top to 0.7
- sphere to 0.2

has a much higher average than the first. But what contradicts me is, a higher average should cause more rolling. And introducing rolling resistance would cause _less_ rolling. But according to above _less_ rolling is _more_ explosive.. Darn.. I am unsure how to proceed now.

**Comment by Waqar144** [comment]
Yes.
And another thing I just noticed is that the explosion is independent of the sphere. Its happens even after the sphere rolled away.
It kind of becomes like a catapult, storing energy and releasing it instantaneously.

**Comment by MrNex** [comment]
That is interesting news. So the problem is not the sphere at all, but the blocks trying to roll. Wow. I'm going to draw some free body diagrams and see if I can pinpoint the problem.

Expect a patch with an attempt at incorporating the `coefficientOfRollingResistance` either later today or tomorrow. Thank you for your help. It is truly invaluable.

**Comment by Waqar144** [comment]
I think printing out the values at different timesteps may help more in pinpointing what's exactly going on.

**Comment by MrNex** [comment]
I plan on writing a debug state which we can attach to specific objects to
print certain values each time step. Something like that has been needed
for a while now.

I have one last thought. Does changing the order we create and add the
bodies to the object manager change the results of the simulation? If this
is the case, rolling resistance will not help us, we need an algorithm for
simulating multiple simultaneous collisions on a single body.

**Comment by Waqar144** [comment]
It shouldn't, theoretically speaking. But, we will need an algorithm for simulating multiple simultaneous collisions.
But this could be solved using multi-threading? no?

**Comment by MrNex** [comment]
No, Multithreading would lead to even more problems (Now there is a possibility one thread can be writing to memory in the middle of another trying to pull the same memory into a register for operations, among other issues).

I'll need to actually test it, but I think the order in which the collisions are solved _might_ make a difference. One part of me says yes, another part says no. There's only one way to find out!

So three things to do:
- Write a state which can be attached to objects to print certain fields each timestep
- Test if the order in which the collisions are solved makes a difference on the results (Specifically changing the order of the floor, bottom box, and top box)
- If all else fails and there are still no leads to follow add rolling resistance.

**Comment by MrNex** [comment]
Here are some updates:
- The debug state was created
- I performed 3 tests and logged the outputs. What I printed was the angular velocity of the bottom block only when it was in contact with >2 things. Here were the results:

(Note: If you are unfamiliar with this format of angular velocity, the direction represents the axis (through the center of mass) around which the block is attempting to rotate _counter clockwise_ and the magnitude represents the amount of the rotation in radians per second.)

Test 1

```
Bottom Block:   [0.000000, 0.000000, 0.000000]
Bottom Block:   [2.847066, 0.000000, 0.000000]
Bottom Block:   [9.069862, 0.000000, 0.000000]
Bottom Block:   [-11.772102, 0.000000, 0.000000]
Bottom Block:   [23.671982, 0.000000, 0.000000]
Bottom Block:   [-53.009605, 0.000000, 0.000000]
```

Test 2

```
Bottom Block:   [0.000000, 0.000000, 0.000000]
Bottom Block:   [2.845908, 0.000000, 0.000000]
Bottom Block:   [-36.696434, 0.000000, 0.000000]
Bottom Block:   [9.226177, 0.000000, 0.000000]
...
Bottom Block:   [10.030186, 0.000000, 0.000000]
...
```

Test 3

```
Bottom Block:   [0.000000, 0.000000, 0.000000]
Bottom Block:   [2.847381, 0.000000, 0.000000]
Bottom Block:   [-35.928951, 0.000000, 0.000000]
Bottom Block:   [-100.891373, 0.000000, 0.000000]
Bottom Block:   [-231.708954, 0.000000, 0.000000]
Bottom Block:   [79.120300, 0.000000, 0.000000]
Bottom Block:   [80.992767, 0.000000, 0.000000]
```

The tests seem consistent with my guess. It should be noted Test 1 was the most explosive, followed by Test 3, and Test 2. Judging by the frequent Back-And-Forth of the sign of the angular velocity before the block goes airborn it appears there are a few collisions where the block tries to rotate one way, but collides with both the floor and the top block (And the sphere) at which point its angular velocity **grows** and switches to the opposite direction, before it is finally set free and the system explodes.

It should also be noted that in tests 3 and (very much) 2, the block was released before it had a chance to explode, and began to slide backward with the top block still there, releasing some energy.

My next test will be to determine if this is related to the order of the solving of the collisions. I will post my findings.

**Comment by MrNex** [comment]
It had been confirmed that the order in which the collisions are solved makes no difference on the results. Although it **seemed** like one order in particular had a less probability of exploding, the claim was not experimented with enough trials. You are right in your call, that is not the issue.

I have one more theory to test, what if we put the heavier block on the bottom..

**Comment by MrNex** [comment]
This issue was fixed with commit a28f39b however a few notes:
- Systems WILL rightfully blow up without giving them physically correct values
- Cubes should **not** be given low rolling resistances
- Near-perfect elastic spheres **should** be given very low rolling resistances (`~ 0.00001f`)
- large stacks of objects still seem to blow up when an object hits them if the objects have non-zero frictional coefficients.

This solution still needs more thorough testing. I am going to close it tentatively.
This solution also presents a new problem, the IsResolutionNeeded function should be performed **after** the decoupling of a collision and the deduction of the collision point. Therefore, it must be altered to be more accurate.
